### PR TITLE
Migration to C++17 code refactor for loops

### PR DESCRIPTION
**Pull request by GermanAizek**
@hrydgard Possible improvement codegeneration by compiler, improvement readability, reduction code size.

### CODE REVIEW

**Comment by hrydgard** [comment]
The build is failing on Windows for some reason. Take a look at the build logs and fix it up please, then I'll take a look.

File: GPU/Common/TextureReplacer.cpp [hunk 1]
```diff
@@ -271,37 +271,37 @@ bool TextureReplacer::LoadIniValues(IniFile &ini, VFSBackend *dir, bool isOverri
...
 			int level = 0;
-			if (sscanf(item.first.c_str(), "%16llx%8x_%d", &key.cachekey, &key.hash, &level) >= 1) {
+			if (sscanf(k.c_str(), "%16llx%8x_%d", &key.cachekey, &key.hash, &level) >= 1) {
 				// We allow empty filenames, to mark textures that we don't want to keep saving.
-				filenameMap[key][level] = item.second;
+				filenameMap[key][level] = v;
 				if (checkFilenames) {
 					// TODO: We should check for the union of these on all platforms, really.
 #if PPSSPP_PLATFORM(WINDOWS)
 					bool bad = item.second.find_first_of("\\ABCDEFGHIJKLMNOPQRSTUVWXYZ:<>|?*") != std::string::npos;
```
Comment (Nemoumbra) [inline GPU/Common/TextureReplacer.cpp:285]: D:\a\ppsspp\ppsspp\GPU\Common\TextureReplacer.cpp(285,17): error C2065: 'item': undeclared identifier

Comment (GermanAizek) [reply GPU/Common/TextureReplacer.cpp:285]: @Nemoumbra I built PPSSPP on Debian, I didn't check it on Windows. I'll fix it now.

Comment (GermanAizek) [reply GPU/Common/TextureReplacer.cpp]: fix it https://github.com/hrydgard/ppsspp/pull/18559/commits/63337c5a53a40af95b114c094550b9eda8dab0cd

File: Common/Serialize/SerializeMap.h [hunk 2]
```diff
@@ -62,9 +62,9 @@ void DoMap(PointerWrap &p, M &x, typename M::mapped_type &default_val) {
 template<class K, class T>
 void Do(PointerWrap &p, std::map<K, T *> &x) {
 	if (p.mode == PointerWrap::MODE_READ) {
-		for (auto it = x.begin(), end = x.end(); it != end; ++it) {
-			if (it->second != nullptr)
-				delete it->second;
+		for (const auto &[_,value] : x) {
```
Comment (hrydgard) [inline Common/Serialize/SerializeMap.h:65]: Wow, this is getting a little to modern for me haha.

File: Common/System/OSD.cpp [hunk 3]
```diff
@@ -258,20 +258,20 @@ void OnScreenDisplay::SetProgressBar(const std::string &id, std::string &&messag
...
+				if (entry.maxValue != 0.0f) {
+                    entry.progress = entry.maxValue;
 				} else {
 					// Fake a full progress
-					iter->minValue = 0;
-					iter->maxValue = 1;
-					iter->progress = 1;
+                    entry.minValue = 0;
+                    entry.maxValue = 1;
+                    entry.progress = 1;
```
Comment (hrydgard) [inline Common/System/OSD.cpp:271]: Tabs are wrong

File: Common/System/OSD.cpp [hunk 4]
```diff
@@ -258,20 +258,20 @@ void OnScreenDisplay::SetProgressBar(const std::string &id, std::string &&messag
...
 				}
 			}
-			iter->endTime = time_now_d() + delay_s + FadeoutTime();
+            entry.endTime = time_now_d() + delay_s + FadeoutTime();
```
Comment (hrydgard) [inline Common/System/OSD.cpp:274]: tabs

File: Common/UI/Root.cpp [hunk 5]
```diff
@@ -353,21 +353,21 @@ static void ProcessHeldKeys(ViewGroup *root) {
...
 			focusMoves.push_back(key.keyCode);
 
 			// Cannot modify the current item when looping over a set, so let's do this instead.
-			HeldKey hk = *iter;
-			heldKeys.erase(hk);
-			hk.triggerTime = now + repeatInterval;
-			heldKeys.insert(hk);
+			HeldKey k = hk;
+			heldKeys.erase(k);
+            k.triggerTime = now + repeatInterval;
```
Comment (hrydgard) [inline Common/UI/Root.cpp:369]: tabs

File: Common/UI/Screen.cpp [hunk 6]
```diff
@@ -149,8 +149,8 @@ void ScreenManager::resized() {
 	std::lock_guard<std::recursive_mutex> guard(inputLock_);
 	// Have to notify the whole stack, otherwise there will be problems when going back
 	// to non-top screens.
-	for (auto iter = stack_.begin(); iter != stack_.end(); ++iter) {
-		iter->screen->resized();
+	for (auto &layer : stack_) {
+        layer.screen->resized();
```
Comment (hrydgard) [inline Common/UI/Screen.cpp:153]: tabs

**Review by hrydgard — CHANGES_REQUESTED** [review]
Please configure tabs in your editor correctly.

File: Common/UI/Screen.cpp (hunk 6 above)
Comment (GermanAizek) [reply Common/UI/Screen.cpp]: fix it https://github.com/hrydgard/ppsspp/pull/18559/commits/37ce2a4de179e6931b8f7e5de4c78728ac0c5c5b

File: Common/UI/Root.cpp (hunk 5 above)
Comment (GermanAizek) [reply Common/UI/Root.cpp]: fix it https://github.com/hrydgard/ppsspp/pull/18559/commits/37ce2a4de179e6931b8f7e5de4c78728ac0c5c5b

File: Common/System/OSD.cpp (hunk 4 above)
Comment (GermanAizek) [reply Common/System/OSD.cpp]: fix it https://github.com/hrydgard/ppsspp/pull/18559/commits/37ce2a4de179e6931b8f7e5de4c78728ac0c5c5b

File: Common/System/OSD.cpp (hunk 3 above)
Comment (GermanAizek) [reply Common/System/OSD.cpp]: fix it https://github.com/hrydgard/ppsspp/pull/18559/commits/37ce2a4de179e6931b8f7e5de4c78728ac0c5c5b

File: Common/Serialize/SerializeMap.h (hunk 2 above)
Comment (GermanAizek) [reply Common/Serialize/SerializeMap.h]: fix it https://github.com/hrydgard/ppsspp/pull/18559/commits/37ce2a4de179e6931b8f7e5de4c78728ac0c5c5b

**Comment by hrydgard** [comment]
There's a conflict now.

**Comment by anr2me** [comment]
You need to choose one version(based on the branch name) of the change to resolve this conflict
image
either removes the code between `<<<<<`...`=====` or the code between `=====`...`>>>>>` (you can see the branch name on the right side of the arrows)

File: Common/Serialize/SerializeList.h [hunk 7]
```diff
@@ -27,9 +27,8 @@ void DoList(PointerWrap &p, std::list<T> &x, T &default_val) {
 	Do(p, list_size);
 	x.resize(list_size, default_val);
 
-	typename std::list<T>::iterator itr, end;
-	for (itr = x.begin(), end = x.end(); itr != end; ++itr)
-		Do(p, *itr);
+	for (T elem : x)
```
Comment (unknownbrackets) [inline Common/Serialize/SerializeList.h:30]: It's actually important that this does not perform a copy, as it did not before.

-[Unknown]

File: Common/Serialize/SerializeMap.h [hunk 8]
```diff
@@ -152,9 +152,9 @@ void Do(PointerWrap &p, std::multimap<K, T> &x) {
 template<class K, class T>
 void Do(PointerWrap &p, std::unordered_multimap<K, T *> &x) {
 	if (p.mode == PointerWrap::MODE_READ) {
-		for (auto it = x.begin(), end = x.end(); it != end; ++it) {
-			if (it->second != nullptr)
-				delete it->second;
+		for (const auto &[_, value] : x) {
+			if (value != nullptr)
+				delete value;
```
Comment (unknownbrackets) [inline Common/Serialize/SerializeMap.h:157]: nit: Personally, I like to always pretend that `delete` mutates a pointer, and avoid `const` for things I'm going to delete.  Semantically, I think it'd be confusing to see a `const int *p` argument to a function which ends up actually deleting that pointer.

Anyway, it sometimes does in fact mutate the object by calling the destructor.

-[Unknown]

File: Common/UI/Root.cpp [hunk 9]
```diff
@@ -353,21 +353,21 @@ static void ProcessHeldKeys(ViewGroup *root) {
...
 			// Cannot modify the current item when looping over a set, so let's do this instead.
-			HeldKey hk = *iter;
-			heldKeys.erase(hk);
-			hk.triggerTime = now + repeatInterval;
-			heldKeys.insert(hk);
+			HeldKey k = hk;
+			heldKeys.erase(k);
+			k.triggerTime = now + repeatInterval;
+			heldKeys.insert(k);
 			goto restart;
```
Comment (unknownbrackets) [inline Common/UI/Root.cpp:371]: I always worry about these goto-restart things with range-fors.  It's less clear what tricks the compiler might do and whether it's safe.  Maybe this should be a `std::remove_if` pattern or something.

-[Unknown]

File: Core/HLE/proAdhocServer.cpp [hunk 10]
```diff
@@ -1376,8 +1376,8 @@ void game_product_override(SceNetAdhocctlProductCode * product)
 				// Destroy Prepare SQL Statement
 				sqlite3_finalize(statement);
 			}*/
-			for (std::vector<db_productid>::iterator it = productids.begin(); it != productids.end(); it++) {
-				if (IsMatch(it->id, productid)) {
+			for (const auto &product: productids) {
```
Comment (unknownbrackets) [inline Core/HLE/proAdhocServer.cpp:1379]: nit:
```suggestion
			for (const auto &product : productids) {
```

-[Unknown]

File: Common/Serialize/SerializeMap.h (hunk 8 above)
Comment (Nemoumbra) [reply Common/Serialize/SerializeMap.h:157]: If I'm not mistaken, delete works fine with the nullptr so the check is superfluous.

**Comment by GermanAizek** [comment]
@hrydgard,
Oh, I thought this PR had been merged for a long time, how do I fix conflicts now?

https://github.com/hrydgard/ppsspp/pull/18559#discussion_r1437989747
I think requirement fixed missing space is not important

**Comment by GermanAizek** [comment]
@hrydgard, and I noticed that in github I can add a commit to PR on Web version, it's convenient. Done.

**Comment by Nemoumbra** [comment]
> Oh, I thought this PR had been merged for a long time
You have not resolved 2 conversations with Unknown regarding your changes, apparently the one in `SerializeList.h` is very important.

File: Common/UI/Root.cpp (hunk 9 above)
Comment (GermanAizek) [reply Common/UI/Root.cpp]: fix it here https://github.com/hrydgard/ppsspp/pull/18559/commits/4ce707797481c4d8682f0ca526fa03c1dcccf368

File: Common/Input/InputState.cpp [hunk 11]
```diff
@@ -44,8 +44,8 @@ static std::unordered_map<InputDeviceID, int> uiFlipAnalogY;
 
 static void AppendKeys(std::vector<InputMapping> &keys, const std::vector<InputMapping> &newKeys) {
 	keys.reserve(newKeys.size());
```
[the remark this answers is not in the document]
Comment (GermanAizek) [reply Common/Input/InputState.cpp]: fix it https://github.com/hrydgard/ppsspp/pull/18559/commits/4ce707797481c4d8682f0ca526fa03c1dcccf368

**Comment by GermanAizek** [comment]
@hrydgard, im fixed all conflicts and add fixes Unknown to commit.

**Comment by hrydgard** [comment]
Replacing with the rebase #19470 .

**Pull request closed without merging**